In the part3 we will construct a P2PKH transaction with sereval inputs and outsputs. We will construct a transaction with 2 inputs and three outputs.
Intersting fact is that that if we spend coins from some transaction we have to spend them all at once, because the difference input - output will be transfered to the miner as reward. If we send bitcoin through Bitcoin Core GUI (as almost all other wallets) this will be handled automatically. The wallet software will create onther address and transfer the difference to it (minus miner fee, for the transaction to be accepted). Anyhow if we manually create raw transaction or even use 'createrawtransaction' command in Bitcoin Core we have to handle this manually. You have probably heard the concept of the change address which comes from this fact
Also the transaction that we can spend are called in Bitcoin UTXO (Unspent Transaction Output)
- It seems that coincurve still somehow produces sometimes different signatures as Bitcoin Core. So the only way is to test my transaction is to sent it 

In [3]:
# lets create 3 new recipient addresses. We will use Bitcoin Core for this, but it could be also done manually as described in the part1
from bitcoinrpc.authproxy import AuthServiceProxy

rpc_user = "pawel"
rpc_password = "qqqwwweee"
rpc_port = "18332"  # Default port for Bitcoin Core RPC is 8332. Use 18332 for testnet.
rpc_connection = AuthServiceProxy(f"http://{rpc_user}:{rpc_password}@localhost:{rpc_port}")

addr1 = rpc_connection.getnewaddress('', 'legacy')
addr2 = rpc_connection.getnewaddress('', 'legacy')
addr3 = rpc_connection.getnewaddress('', 'legacy')

print(addr1)
print(addr2)
print(addr3)

mnNSXkYQiFz73Hi7KnDYgNCzqvpKdVa355
mqoZ1kdfmomVKMgFvZmBTQ3e2dRcho4N2Z
mhSMUAMQGGS9WRWx3XcKhFphhLmvuipqqo


mnNSXkYQiFz73Hi7KnDYgNCzqvpKdVa355
mqoZ1kdfmomVKMgFvZmBTQ3e2dRcho4N2Z
mhSMUAMQGGS9WRWx3XcKhFphhLmvuipqqo


Now we have generated 3 recipient addresses
mnNSXkYQiFz73Hi7KnDYgNCzqvpKdVa355
mqoZ1kdfmomVKMgFvZmBTQ3e2dRcho4N2Z
mhSMUAMQGGS9WRWx3XcKhFphhLmvuipqqo

Lets look agaim at the structure of raw transaction
rawTx_toSign = (
    version             # 4 bytes: Transaction version number
    + tx_in_count      # Variable length: Number of transaction inputs
    + txid             # 32 bytes: Transaction ID of the UTXO being spent
    + vout             # 4 bytes: Index of the UTXO in the transaction
    + scriptSig_len_t   # Variable length: Length of the scriptSig
    + scriptSig_t      # Variable length: Script that provides data to satisfy the UTXO's locking script
    + sequence         # 4 bytes: Sequence number
    + tx_out_count     # Variable length: Number of transaction outputs
    + value            # 8 bytes: Amount of satoshis to send to the recipient
    + scriptPubKey_len # Variable length: Length of the scriptPubKey
    + scriptPubKey     # Variable length: Locking script specifying conditions to spend the output
    + locktime         # 4 bytes: Time or block number when the transaction becomes valid
    + sighash_type     # 4 bytes: Signature hash type (only if signing)
)

As explained in the part2 before we have real signature in the P2PKH transaction we put there the locking condition of the input transaction(pubKey script) in the ScripitSig part (thats why i put *_t in the name to indicate thta it is temporary). After we sign the transaction we will put the real signature in the SigScript part (as I said before it is the same what Bitcoin Core software is also doing to check the signature). 

So lets see which parts are missing, derive them and construct the raw transaction

- TXID
- Vout
- ScriptSig
- Value
- ScriptPubKey

In [12]:
# lets at first define the values for the known variables. 
# Remember that we are constructiong a P2PKH transaction with 2 inputs and 3 outputs
version = '02000000'
tx_in_count = '02'
sequence = 'fdffffff'
tx_out_count = '03'
locktime = '00000000'
sig_hash_type = '01000000'

In [7]:
# lets see which transaction we can spend (if you have less then 2 create a new one)
rpc_user = "pawel"
rpc_password = "qqqwwweee"
rpc_port = "18332"  # Default port for Bitcoin Core RPC is 8332. Use 18332 for testnet.
rpc_connection = AuthServiceProxy(f"http://{rpc_user}:{rpc_password}@localhost:{rpc_port}")

utxos = rpc_connection.listunspent()

for utxo in utxos:
    print(f"Address: {utxo['address']}")
    print(f"Transaction ID: {utxo['txid']}")
    print(f"Output Index: {utxo['vout']}")
    print(f"Script Public Key: {utxo['scriptPubKey']}")
    print(f"Value: {utxo['amount']} BTC")
    print("-------------------------------")

Address: mnPXsbCFtQbLrpAwQUDETcpUApDcjamSmB
Transaction ID: ee3175c87c04a0ec16bf683d2f733289a60b9f8aabf932afd673b792c979b913
Output Index: 2
Script Public Key: 76a9144b61951cee51ecc3c23956b828ebe84068542f2f88ac
Value: 0.00900000 BTC
-------------------------------
Address: mwQ2jDbA6Yzt35r4pcRuCViAwzyyySC4oK
Transaction ID: bcffee2198135c5d6be186aab42d261977bbd2d3fe4d9f0cb933a35b762507a3
Output Index: 0
Script Public Key: 76a914ae32cd505abe9338e792df038d288485c0b601d788ac
Value: 0.00590000 BTC
-------------------------------
Address: n39Zw3LQMMWWKqW4aeowTQM3pgWVvE5uFt
Transaction ID: bcffee2198135c5d6be186aab42d261977bbd2d3fe4d9f0cb933a35b762507a3
Output Index: 1
Script Public Key: 76a914ed475d6cf80598ffbb2d4af64f0421cf553952e788ac
Value: 0.00590000 BTC
-------------------------------
Address: mmgfR5aXVMYJYAJSo8hJbFQzkamYH6Fxs3
Transaction ID: bcffee2198135c5d6be186aab42d261977bbd2d3fe4d9f0cb933a35b762507a3
Output Index: 2
Script Public Key: 76a91443a6d315d938d80025b86c211edc1b72aef2540

Lets spend this 2 UTXOs. The total balance is 0.009 + 0.0059 = 0.0149 BTC.
Lets divide it into 3 outputs evenly. So each output will be 0.0049 btc

Address: mnPXsbCFtQbLrpAwQUDETcpUApDcjamSmB
Transaction ID: ee3175c87c04a0ec16bf683d2f733289a60b9f8aabf932afd673b792c979b913
Output Index: 2
Script Public Key: 76a9144b61951cee51ecc3c23956b828ebe84068542f2f88ac
Value: 0.00900000 BTC
Pubkey: 02ba91253c14afa8cb4ce42ef90353aa60b7cff6497e7a3173198262e5423bbea4

Address: mwQ2jDbA6Yzt35r4pcRuCViAwzyyySC4oK
Transaction ID: bcffee2198135c5d6be186aab42d261977bbd2d3fe4d9f0cb933a35b762507a3
Output Index: 0
Script Public Key: 76a914ae32cd505abe9338e792df038d288485c0b601d788ac
Pubkey: 026b86df5a6e11da7500b0da4749deb77248a69b6f728be5ff2fb2d740fb418282



Now we just have to construct the unlocking conditions (scriptPubkey) for our 3 addresses.
As already explained in the part2 the scriptPubKey for P2PKH is:
OP_DUP OP_HASH160 <PubKeyHash> OP_EQUALVERIFY OP_CHECKSIG

where PubKeyHash is the hash of the public key of the recipient address as hex string. So we have to convert back the recipipient address from base58 back to hex and remove version byte and the checksum. Then we add the OP codes as shown above.

mnNSXkYQiFz73Hi7KnDYgNCzqvpKdVa355
mqoZ1kdfmomVKMgFvZmBTQ3e2dRcho4N2Z
mhSMUAMQGGS9WRWx3XcKhFphhLmvuipqqo

In [30]:
import base58

def address_to_scriptpubkey(address: str) -> str:
    """Convert a P2PKH address to its corresponding scriptPubKey."""
    # Decode the address from Base58
    decoded = base58.b58decode(address)
    
    # Extract the 20-byte PKH (drop the first byte and last 4 bytes)
    pkh = decoded[1:-4]
    
    # Construct the scriptPubKey: OP_DUP OP_HASH160 <pkh> OP_EQUALVERIFY OP_CHECKSIG
    scriptpubkey = '76a914' + pkh.hex() + '88ac'
    
    return scriptpubkey

scriptPubkey1 = address_to_scriptpubkey(addr1)
scriptPubkey2 = address_to_scriptpubkey(addr2)
scriptPubkey3 = address_to_scriptpubkey(addr3)

scriptPubkey1_len = bytes.fromhex(hex(len(scriptPubkey1)//2)[2:]).hex()
scriptPubkey2_len = bytes.fromhex(hex(len(scriptPubkey2)//2)[2:]).hex()
scriptPubkey3_len = bytes.fromhex(hex(len(scriptPubkey3)//2)[2:]).hex()

pubkey1 = '02ba91253c14afa8cb4ce42ef90353aa60b7cff6497e7a3173198262e5423bbea4'
pubkey2 = '026b86df5a6e11da7500b0da4749deb77248a69b6f728be5ff2fb2d740fb418282'


print(scriptPubkey1)
print(scriptPubkey2)
print(scriptPubkey3)

76a9144b2cb50a9a82eb15710b9b048c9334dadf9ecd8788ac
76a91470d4e52849996467ee3e58dace3e90d5ce569e5988ac
76a91415118444824e77374c3245f08686aa9b91f97bbd88ac


Now we have all the iinformation and can start to construct the raw transaction for signing

- Copy here all the infos from RPC call

In [18]:
# lets define all other variables
tx_hash_1 = 'ee3175c87c04a0ec16bf683d2f733289a60b9f8aabf932afd673b792c979b913'
TX_ID_1 = txid = bytes.fromhex(tx_hash_1)[::-1].hex()

tx_hash_2 = 'bcffee2198135c5d6be186aab42d261977bbd2d3fe4d9f0cb933a35b762507a3'
TX_ID_2 = txid = bytes.fromhex(tx_hash_2)[::-1].hex()

v_out_1 = '02000000'
v_out_2 = '00000000'


# at the moment we have 0.0149 and 0.0122 BTC on the TX_ID_1 and TX_ID_2 respectively
# our ouputsum has to be lower than the inputsum (we also have to pay the fee)
# lets send 0.009 BTC to addr1, 0.009 BTC to addr2 and 0.009 BTC to addr3 and 0.001 BTC to the miner



# the temp ScriptSig is the scriptpubkey of the previous transaction. which we got from the RPC
scriptSig1_t = '76a9144b61951cee51ecc3c23956b828ebe84068542f2f88ac'
scriptSig2_t = '76a914ae32cd505abe9338e792df038d288485c0b601d788ac'

scriptSig1_len_t = bytes.fromhex(hex(len(scriptSig1_t)//2)[2:]).hex()
scriptSig2_len_t = bytes.fromhex(hex(len(scriptSig2_t)//2)[2:]).hex()

In [19]:
def btc_to_satoshis_hex(btc_value: float) -> str:
    """Convert BTC to its 8-byte little-endian hex representation."""
    # Convert BTC to satoshi (1 BTC = 100,000,000 satoshis)
    satoshis = round(btc_value * 100_000_000)

    # Convert to 8-byte little endian format and return its hexadecimal representation, 
    # ensuring it's always padded to 16 characters
    return satoshis.to_bytes(8, 'little').hex().rjust(16, '0')

# Example:
btc_value1 = 0.0049
btc_value2 = 0.0049
btc_value3 = 0.0049

value1 = btc_to_satoshis_hex(btc_value1)
value2 = btc_to_satoshis_hex(btc_value2)
value3 = btc_to_satoshis_hex(btc_value3)

print(value1)


107a070000000000


lets agian see the structure of the bitcoin transaction with several inputs and outputs

Version (4 bytes)
Input Count (varint)
    Input 1:
    txid (32 bytes)
    vout (4 bytes)
    scriptSig length (varint)
    scriptSig (variable length, typically empty if unsigned)
    sequence (4 bytes)
Input 2 (repeat structure of Input 1)
Output Count (varint)
Output 1:
    value (8 bytes)
    scriptPubKey length (varint)
    scriptPubKey (variable length)
Output 2 (repeat structure of Output 1)
Output 3 (repeat structure of Output 1)
Locktime (4 bytes)
Sighash type (4 bytes: only if signing)


- During signing ofinputs we have to set temprorary Scriptsig to the Scriptpubke of the previous tx for the TX_ID we are signing AND ALL OTHER TEMP. SCRIPTSIG TO '00'

- Another point is that we got public key from bitcoin wallet, but it is just derivable from the private key as demontsrated in the part2 (it is just corresponding point on the elliptic curve)

In [22]:
# now we can create the raw trasnaction for signing with the first key
rawTx_toSign1 = (
    version
    + tx_in_count  # Now should represent 2 inputs
    + TX_ID_1
    + v_out_1
    + scriptSig1_len_t
    + scriptSig1_t
    + sequence
    + TX_ID_2
    + v_out_2
    + '00'
    #+ scriptSig2_len_t
    #+ scriptSig2_t
    + sequence
    + tx_out_count  # Now should represent 3 outputs
    + value1
    + scriptPubkey1_len
    + scriptPubkey1
    + value2
    + scriptPubkey2_len
    + scriptPubkey2
    + value3
    + scriptPubkey3_len
    + scriptPubkey3
    + locktime
    + sig_hash_type  # Only if signing
)

print(rawTx_toSign1)

020000000213b979c992b773d6af32f9ab8a9f0ba68932732f3d68bf16eca0047cc87531ee020000001976a9144b61951cee51ecc3c23956b828ebe84068542f2f88acfdffffffa30725765ba333b90c9f4dfed3d2bb7719262db4aa86e16b5d5c139821eeffbc0000000000fdffffff03107a0700000000001976a9144b2cb50a9a82eb15710b9b048c9334dadf9ecd8788ac107a0700000000001976a91470d4e52849996467ee3e58dace3e90d5ce569e5988ac107a0700000000001976a91415118444824e77374c3245f08686aa9b91f97bbd88ac0000000001000000


In [24]:
# now we can create the raw trasnaction for signing with the second key
rawTx_toSign2 = (
    version
    + tx_in_count  # Now should represent 2 inputs
    + TX_ID_1
    + v_out_1
    + '00'
   # + scriptSig1_len_t
   # + scriptSig1_t
    + sequence
    + TX_ID_2
    + v_out_2
    + scriptSig2_len_t
    + scriptSig2_t
    + sequence
    + tx_out_count  # Now should represent 3 outputs
    + value1
    + scriptPubkey1_len
    + scriptPubkey1
    + value2
    + scriptPubkey2_len
    + scriptPubkey2
    + value3
    + scriptPubkey3_len
    + scriptPubkey3
    + locktime
    + sig_hash_type  # Only if signing
)

print(rawTx_toSign2)

020000000213b979c992b773d6af32f9ab8a9f0ba68932732f3d68bf16eca0047cc87531ee0200000000fdffffffa30725765ba333b90c9f4dfed3d2bb7719262db4aa86e16b5d5c139821eeffbc000000001976a914ae32cd505abe9338e792df038d288485c0b601d788acfdffffff03107a0700000000001976a9144b2cb50a9a82eb15710b9b048c9334dadf9ecd8788ac107a0700000000001976a91470d4e52849996467ee3e58dace3e90d5ce569e5988ac107a0700000000001976a91415118444824e77374c3245f08686aa9b91f97bbd88ac0000000001000000


020000000213b979c992b773d6af32f9ab8a9f0ba68932732f3d68bf16eca0047cc87531ee020000001976a9144b61951cee51ecc3c23956b828ebe84068542f2f88acfdffffffa30725765ba333b90c9f4dfed3d2bb7719262db4aa86e16b5d5c139821eeffbc0000000000fdffffff03107a0700000000001976a9144b2cb50a9a82eb15710b9b048c9334dadf9ecd8788ac107a0700000000001976a91470d4e52849996467ee3e58dace3e90d5ce569e5988ac107a0700000000001976a91415118444824e77374c3245f08686aa9b91f97bbd88ac0000000001000000

020000000213b979c992b773d6af32f9ab8a9f0ba68932732f3d68bf16eca0047cc87531ee0200000000fdffffffa30725765ba333b90c9f4dfed3d2bb7719262db4aa86e16b5d5c139821eeffbc000000001976a914ae32cd505abe9338e792df038d288485c0b601d788acfdffffff03107a0700000000001976a9144b2cb50a9a82eb15710b9b048c9334dadf9ecd8788ac107a0700000000001976a91470d4e52849996467ee3e58dace3e90d5ce569e5988ac107a0700000000001976a91415118444824e77374c3245f08686aa9b91f97bbd88ac0000000001000000

In [15]:
# before signing the transaction we shall get private keys for the inputs 
# or we already have them if we created them manually as described in the part1

# lets get the private keys for the inputs
rpc_user = "pawel"
rpc_password = "qqqwwweee"
rpc_port = "18332"  # Default port for Bitcoin Core RPC is 8332. Use 18332 for testnet.
rpc_connection = AuthServiceProxy(f"http://{rpc_user}:{rpc_password}@localhost:{rpc_port}")


in_addr1 = 'mnPXsbCFtQbLrpAwQUDETcpUApDcjamSmB'
privKey_WIF_1 = rpc_connection.dumpprivkey(in_addr1)

in_addr2 = 'mwQ2jDbA6Yzt35r4pcRuCViAwzyyySC4oK'
privKey_WIF_2 = rpc_connection.dumpprivkey(in_addr2)

print(privKey_WIF_1)
print(privKey_WIF_2)



cUk2wUnGbnkV87Q24v3mRuhJMMdtN9caMHQdydQvmRM8BPrQxFEd
cUp7j8Fa6rbn9xo1FYiu4qeyNW3Un6r9GKpKskd4iYtwSPXuHo67


cUk2wUnGbnkV87Q24v3mRuhJMMdtN9caMHQdydQvmRM8BPrQxFEd
cUp7j8Fa6rbn9xo1FYiu4qeyNW3Un6r9GKpKskd4iYtwSPXuHo67

In [16]:
# now we have to decode the private keys from WIF format to hex format
import base58
import hashlib

def wif_to_private_key_hex(privKeyWIF: str) -> str:
    """
    Convert a WIF (Wallet Import Format) private key to its hexadecimal representation.
    
    Args:
    - privKeyWIF (str): WIF encoded private key string.
    
    Returns:
    - str: Hexadecimal representation of the private key.
    """
    # 1. Decode the WIF format using Base58
    decoded = base58.b58decode(privKeyWIF)

    # 2. Strip and validate the checksum
    extended_key = decoded[:-4]  # Strip the checksum
    checksum = decoded[-4:]

    # Verify the checksum
    if checksum != hashlib.sha256(hashlib.sha256(extended_key).digest()).digest()[:4]:
        raise ValueError("Invalid WIF checksum!")

    # 3. Remove the version byte
    extended_key = extended_key[1:]

    # 4. Check and remove the 01 byte for compressed pubKey (if present)
    if extended_key[-1] == 1:  # Checking if the last byte is 0x01
        privKey_hex = extended_key[:-1].hex()  # Removing the 01 byte
    else:
        privKey_hex = extended_key.hex()
    
    return privKey_hex

# Usage
priv_key_hex_1 = wif_to_private_key_hex(privKey_WIF_1)
priv_key_hex_2 = wif_to_private_key_hex(privKey_WIF_2)

print(priv_key_hex_1)
print(priv_key_hex_2)



d5a9aeecf7cd432b3fc968f9300da8a70efd464d6374aec3adab2a2fd4d25b82
d7c3564fd38bcd83ca63cc95f844f856875305aea87f44c57439e008619591a5


d5a9aeecf7cd432b3fc968f9300da8a70efd464d6374aec3adab2a2fd4d25b82
d7c3564fd38bcd83ca63cc95f844f856875305aea87f44c57439e008619591a5

In [25]:
# lets sign the transaction
# at first we have to hash the transaction twice (double sha256)
# dont forget that for P2PKH we have to add the public key to the signature
# also remember that we have to sign the transaction twice (once for each input)
# dont forget to add the sighash type at the end of the signature (01)

from coincurve import PrivateKey

def sign_transaction(privKey_hex, rawTx_toSign):
    """
    Sign a raw transaction using the provided private key.
    
    Args:
    - privKey_hex (str): The private key in hexadecimal format.
    - rawTx_toSign_hash (bytes): The hash of the raw transaction to be signed.
    
    Returns:
    - str: The signed transaction (scriptSig) in hexadecimal format.
    """
    # Hash the raw transaction twice
    rawTx_toSign_hash = hashlib.sha256(hashlib.sha256(bytes.fromhex(rawTx_toSign)).digest()).digest()

    # Convert the hex private key to a coincurve PrivateKey object
    privkey = PrivateKey(bytes.fromhex(privKey_hex))
    
    # Sign the transaction hash deterministically
    signature = privkey.sign(rawTx_toSign_hash, hasher=None)
    
    # To signal that we used SigHash all, add '01' at the end
    signature += bytes.fromhex('01')
    
    # Convert the signature to its hex representation
    sig = signature.hex()
    
    return sig

# Usage example:

Sig1 = sign_transaction(priv_key_hex_1, rawTx_toSign1)
Sig2 = sign_transaction(priv_key_hex_2, rawTx_toSign2)


print(Sig1)
print(Sig2)





3044022049032bfb2871d52d9cf0e3527fc70cf6b581125c04bacf92b667066554db789102207cf1780102e78665682231b083f0ca1b4428ea19b70e3e182fb0661266e8bcc001
3045022100bbdc9ad6a867ff53a366cab7f92c5b7daa849e0b443300492b54eb5206f031b802204eee44e88dd6ef37ba5d6089ce5546dbc96eee1a4fc722b85cd40bceade3f98301


3044022049032bfb2871d52d9cf0e3527fc70cf6b581125c04bacf92b667066554db789102207cf1780102e78665682231b083f0ca1b4428ea19b70e3e182fb0661266e8bcc001

3045022100bbdc9ad6a867ff53a366cab7f92c5b7daa849e0b443300492b54eb5206f031b802204eee44e88dd6ef37ba5d6089ce5546dbc96eee1a4fc722b85cd40bceade3f98301

Now lets build the whole transaction. Lets not forget to append the public key to the signatures as required by P2PKH transaction

In [49]:
# so complete scriptSig for the first input is:
# the number 2 assumes here that sig and pubkey are each 1 bytes long (which is not always the case)

len_Sig1 = bytes.fromhex(hex(len(Sig1)//2)[2:]).hex()
len_pubkey1 = bytes.fromhex(hex(len(pubkey1)//2)[2:]).hex()
total_len2 = hex((int(len_Sig1, 16) + int(len_pubkey1, 16) + 2))

scriptSig1 = (total_len2[2:] +len_Sig1 + Sig1 +  len_pubkey1 + pubkey1)

len_Sig2 = bytes.fromhex(hex(len(Sig2)//2)[2:]).hex()
len_pubkey2 = bytes.fromhex(hex(len(pubkey2)//2)[2:]).hex()
total_len2 = hex((int(len_Sig2, 16) + int(len_pubkey2, 16) + 2))

scriptSig2 = (total_len2[2:] +len_Sig2 + Sig2 +  len_pubkey2 + pubkey2)


In [53]:
# now we can build the complete raw transaction
# now we can create the raw trasnaction for signing with the second key
rawTx = (
    version
    + tx_in_count  # Now should represent 2 inputs
    + TX_ID_1
    + v_out_1
    + scriptSig1
    + sequence
    + TX_ID_2
    + v_out_2
    + scriptSig2
    + sequence
    + tx_out_count  # Now should represent 3 outputs
    + value1
    + scriptPubkey1_len
    + scriptPubkey1
    + value2
    + scriptPubkey2_len
    + scriptPubkey2
    + value3
    + scriptPubkey3_len
    + scriptPubkey3
    + locktime
)

print(rawTx)

020000000213b979c992b773d6af32f9ab8a9f0ba68932732f3d68bf16eca0047cc87531ee020000006a473044022049032bfb2871d52d9cf0e3527fc70cf6b581125c04bacf92b667066554db789102207cf1780102e78665682231b083f0ca1b4428ea19b70e3e182fb0661266e8bcc0012102ba91253c14afa8cb4ce42ef90353aa60b7cff6497e7a3173198262e5423bbea4fdffffffa30725765ba333b90c9f4dfed3d2bb7719262db4aa86e16b5d5c139821eeffbc000000006b483045022100bbdc9ad6a867ff53a366cab7f92c5b7daa849e0b443300492b54eb5206f031b802204eee44e88dd6ef37ba5d6089ce5546dbc96eee1a4fc722b85cd40bceade3f9830121026b86df5a6e11da7500b0da4749deb77248a69b6f728be5ff2fb2d740fb418282fdffffff03107a0700000000001976a9144b2cb50a9a82eb15710b9b048c9334dadf9ecd8788ac107a0700000000001976a91470d4e52849996467ee3e58dace3e90d5ce569e5988ac107a0700000000001976a91415118444824e77374c3245f08686aa9b91f97bbd88ac00000000


020000000213b979c992b773d6af32f9ab8a9f0ba68932732f3d68bf16eca0047cc87531ee020000006a473044022049032bfb2871d52d9cf0e3527fc70cf6b581125c04bacf92b667066554db789102207cf1780102e78665682231b083f0ca1b4428ea19b70e3e182fb0661266e8bcc0012102ba91253c14afa8cb4ce42ef90353aa60b7cff6497e7a3173198262e5423bbea4fdffffffa30725765ba333b90c9f4dfed3d2bb7719262db4aa86e16b5d5c139821eeffbc000000006b483045022100bbdc9ad6a867ff53a366cab7f92c5b7daa849e0b443300492b54eb5206f031b802204eee44e88dd6ef37ba5d6089ce5546dbc96eee1a4fc722b85cd40bceade3f9830121026b86df5a6e11da7500b0da4749deb77248a69b6f728be5ff2fb2d740fb418282fdffffff03107a0700000000001976a9144b2cb50a9a82eb15710b9b048c9334dadf9ecd8788ac107a0700000000001976a91470d4e52849996467ee3e58dace3e90d5ce569e5988ac107a0700000000001976a91415118444824e77374c3245f08686aa9b91f97bbd88ac00000000

In [ ]:
# Now we can send the transaction to the network
from bitcoinrpc.authproxy import AuthServiceProxy
rpc_user = "pawel"
rpc_password = "qqqwwweee"
rpc_port = "18332"  # Default port for Bitcoin Core RPC is 8332. Use 18332 for testnet.
rpc_connection = AuthServiceProxy(f"http://{rpc_user}:{rpc_password}@localhost:{rpc_port}")


sent = rpc_connection.sendrawtransaction(rawTx)
#print('TX_ID: ', sent)

- WELL DONE